### DISCUSSION TITLE
Is dead code elimination possible?

### DISCUSSION THREAD

**Discussion by zetachang** [opal/opal#2208] [General]
My initial attempt is to pipe the JS code generated to other compiler. Below is a basic example. 

```ruby
p = true
if p
  puts "live code"
else
  puts "dead code"
end
```

However, [google's closure compiler](https://developers.google.com/closure/compiler/) couldn't eliminate the dead code. It seems like the comparison with `Opal.nil` generated at the if statement break the optimization.

**Comment by zetachang** [comment]
I think my problem might be solved by writing a custom rewriter like https://github.com/opal/opal/blob/master/lib/opal/rewriters/opal_engine_check.rb

The motivation behind this is to strip out some code only used in development. (e.g. warning / validation) which might bloat the built size or affect performance.

**Comment by elia** [comment]
That would certainly do the job. 

I did some experiments some time ago tracking all method calls and skipping the defs for unused methods but didn't get to significant results. Probably the tracking system wasn't up to the task. But I still think that would be of great value if it cuts all the cool stuff from te corelib that it's good to have at hand but are not used every time.

**Comment by zetachang** [comment]
Just out of curiosity, in the `OpalEngineCheck` rewriter, could we savely inline the true body or false body without the if condition test?

**Comment by iliabylich** [comment]
@zetachang In general case the answer is no. We can optimize only some easiest cases like `if true` or `if RUBY_ENGINE == 'opal'`.All other optimizations are impossible due to the dynamicity of the language (stuff like `eval` and `send`)

> Just out of curiosity, in the OpalEngineCheck rewriter, could we savely inline the true body or false body without the if condition test?

Yes, it definitely can be optimized.

> My initial attempt is to pipe the JS code generated to other compiler. Below is a basic example.


``` ruby
p = true
if p
  puts "live code"
else
  puts "dead code"
end
```
It can be optimized with a few assumptions.
1. The value that you check in the `if` must be constant (local variable defined in the current scope with a constant value). Methods are not allowed since we can easily override them in rutime.
2. `if` must be called right after assignment to make sure that there's no `eval` that may change it (`eval("p = false")`)

More compelx cases have much more requirements :smile:

**Comment by zetachang** [comment]
Facebook just released [Prepack](https://prepack.io/), a tool which optimizes code by running (partially) instead of doing static analysis only as Closure compiler.

Code like below (slightly modified from output of Opal compiler) could actually be optimized.

```js
Opal = {};
Opal.top = {};
Opal.top.$puts = function(msg){ console.log(msg); };

/* Generated by Opal 0.10.3 */
(function(Opal) {
  var self = Opal.top, $scope = Opal, nil = Opal.nil, $breaker = Opal.breaker, $slice = Opal.slice, p = nil;

  p = true;
  if (p !== false && p !== nil && p != null) {
    return self.$puts("live code")
    } else {
    return self.$puts("dead code")
  };
})(Opal);
```

becomes

```js
(function () {
  var _$0 = this;

  function _2(msg) {
    _$0.console.log(msg);
  }

  Opal = {
    top: {
      $puts: _2
    }
  };
  console.log("live code");
}).call(this);
```

**Comment by iliabylich** [comment]
@zetachang I'm not sure that Prepack is quite stable right now. At least, it doesn't work well for opal. I've tried to compile an opal + `puts 123` using prepack and after small manipulations I've got the following:
1. Pure opal:
``` sh
➜  prepack git:(master) ✗ wc /tmp/opaltest.js
   21152   84709  782893 /tmp/opaltest.js
➜  prepack git:(master) ✗ repeat 5 time node /tmp/opaltest.js
123
node /tmp/opaltest.js  0.26s user 0.03s system 98% cpu 0.294 total
123
node /tmp/opaltest.js  0.24s user 0.03s system 97% cpu 0.279 total
123
node /tmp/opaltest.js  0.25s user 0.03s system 97% cpu 0.287 total
123
node /tmp/opaltest.js  0.25s user 0.03s system 97% cpu 0.283 total
123
node /tmp/opaltest.js  0.25s user 0.03s system 97% cpu 0.281 total
```
2. Opal compiled using Prepack
``` sh
➜  prepack git:(master) ✗ wc /tmp/opaltest.js-processed.js
   78219  242539 2124511 /tmp/opaltest.js-processed.js
➜  prepack git:(master) ✗ repeat 5 time node /tmp/opaltest.js-processed.js
123

node /tmp/opaltest.js-processed.js  0.24s user 0.04s system 96% cpu 0.292 total
123

node /tmp/opaltest.js-processed.js  0.22s user 0.04s system 96% cpu 0.265 total
123

node /tmp/opaltest.js-processed.js  0.22s user 0.04s system 96% cpu 0.273 total
123

node /tmp/opaltest.js-processed.js  0.23s user 0.04s system 97% cpu 0.277 total
123

node /tmp/opaltest.js-processed.js  0.22s user 0.04s system 96% cpu 0.273 total
```
Size of the source code got tripled, performance is pretty much the same. Also for some reason it rewrites an AST with a weird bug, there's an extra empty line after `puts` 😄 

I'll do more benchmarks with arithmetics and classes/methods.

Also I have to mention that I can't compile rubyspec test suite atm, getting segfault when trying to compile opal + parser + mspec + one small file from rubyspec.

**Comment by iliabylich** [comment]
Prepack still fails to handle opal's corelib:
```
$ opal -c 'puts 1' > tmp/test.js
$ node tmp/test.js
1
$ node --stack-size=100000 node_modules/prepack/bin/prepack.js tmp/test.js --out tmp/test-prepacked.js
Prepacked source code written to tmp/test-prepacked.js.
$ node tmp/test-prepacked.js
tmp/test-prepacked.js:69
  var _$3 = _$2.prototype;
                ^

TypeError: Cannot read property 'prototype' of undefined
```

From what I see in the compiled `test-prepacked.js` the optimization doesn't work well. For example, my favorite method `Number#step` is still there, but I'm 100% sure that corelib doesn't invoke it.

By the way, the size of the "optimized" version is still huge:
```

2.1M Jan  5 15:28 test-prepacked.js
753K Jan  5 15:24 test.js

**Comment by hmdne** [comment]
I tested prepack today on my partial corelib and... it worked. It generated a working code that worked in browser as advertised. On Node... well, it generated a code that is an equivalent to:

```js
(function () {
  this.Object.defineProperty;
}).call(this);
```

And it caused Node to crash with a similar error to yours (which means that I bet your code would run in a browser and if you wanted to target Node, you would need to add a switch to the prepack executable). It IS larger:

```
-rw-rw-r--. 1 user user 503K Nov 21 02:53 minilib.js
-rw-rw-r--. 1 user user 1.1M Nov 21 02:54 minilib.ppk.js
```

I guess that is due to some loop unwinding or meta-programming being executed.

> For example, my favorite method Number#step is still there

It obviously is and that is an intended behaviour. Prepack aims to preserve the environment as much as possible. If Opal didn't modify the global variables (and it needs to) it could try to eliminate some unused functions, but a library makes all functions visible. And even if it didn't, let's not forget that Opal modifies the prototypes. Those won't be eliminated.
